### DISCUSSION TITLE
Battery Uncalibrated, No way to enable AXP223 calibration mode.

### DISCUSSION THREAD

**Discussion by DevL0rd** [OnionUI/Onion#1711] [General]
### Checklist

- [X] I am reporting a bug in the Onion OS - Any issues with the closed source frontend ([MainUI](https://github.com/OnionUI/MainUI-issues)), or [RetroArch](https://github.com/libretro/RetroArch) or any of the cores should be directed to their respective repositories
- [X] I have reviewed the [Onion Docs](https://onionui.github.io/docs) and did not find the answer
- [X] I have searched the existing (open and closed) issues to ensure I do not log a duplicate

### Onions OS Version

4.3.1-1

### Miyoo Firmware Version

202306282128

### Provide a clear and concise description of the issue

My understanding is that onion os is getting battery percent readings directly from the AXP223 chippies.

Unfortunately it's the most in-accurate reading I have ever seen.

Additionally after reading over the document for this controller, there is in fact a calibration mode that can be set.

I believe calibrating it may resolve the issue with crazy battery readings.

May there possibly be a way to toggle the batteries calibration mode from the advanced tools?

My understanding is that you reset it in some way and charge and discharge it to get a more accurate calibration.

### Steps to Reproduce

Please provide detailed steps that will help us replicate and investigate your issue

### Provide any additional context or information that may help us investigate your issue

Use the onion os.
Cry at battery reading inaccuracies.

**Comment by thomasklemenc** [comment]
Is this on the Miyoo Mini or Miyoo Mini plus? Only the plus uses the AXP223, the regular mini has no such chip

**Comment by DevL0rd** [comment]
> Is this on the Miyoo Mini or Miyoo Mini plus? Only the plus uses the AXP223, the regular mini has no such chip

Oh really? Interesting.

Perhaps I can correct my own voltage curve to something more linear.

Can you give me some guidance as to how the percentage is calculated at the moment and where in code that may be?

I assume cause of the mostly constant voltage tho that may be impossible but I'd like to experiment.

**Comment by thomasklemenc** [comment]
Basically on the Miyoo Mini, all we have is the battery voltage. it is read and converted to a percentage by batmon [here](https://github.com/OnionUI/Onion/blob/58916b8bd3308853918991e5bfa25911200a72e0/src/batmon/batmon.c#L84) and then written to /tmp/.axp_result [here](https://github.com/OnionUI/Onion/blob/58916b8bd3308853918991e5bfa25911200a72e0/src/batmon/batmon.c#L325)

MainUI (the closed source frontend) reads the value from that file and displays it accordingly.

Because we only have battery voltage, the percentage can be inaccurate as the voltage dips with higher current draw (high CPU load)

**Reply by DevL0rd** [reply]
> Basically on the Miyoo Mini, all we have is the battery voltage. it is read and converted to a percentage by batmon [here](https://github.com/OnionUI/Onion/blob/58916b8bd3308853918991e5bfa25911200a72e0/src/batmon/batmon.c#L84) and then written to /tmp/.axp_result [here](https://github.com/OnionUI/Onion/blob/58916b8bd3308853918991e5bfa25911200a72e0/src/batmon/batmon.c#L325)
> 
> MainUI (the closed source frontend) reads the value from that file and displays it accordingly.
> 
> Because we only have battery voltage, the percentage can be inaccurate as the voltage dips with higher current draw (high CPU load)

True. I assumed so.

I could calculate a voltage curve I guess for the CPU and correct the current voltage with it for the battery calculation. But I assume the curve there also changes with remaining capacity..

It is a very dynamic system haha but I'm sure I can improve accuracy somehow.

You know if anything in the system I can call to get current power draw so I can include it in the calculation?

Never touched this kind of stuff so this will be my first go at it

**Reply by thomasklemenc** [reply]
I believe there is no way to get the current power draw on the MM

**Reply by DevL0rd** [reply]
> I believe there is no way to get the current power draw on the MM

Mm perhaps I can use a voltmeter and write a program to use 0 - 100% CPU and also change the screen brightness to get precalculates values of screen brightness and CPU usage and the effect on voltage.

All these things together I may be able to improve accuracy.

Anyways thanks. I'm gonna experiment now!

**Comment by DevL0rd** [comment]
You know what. I could just record the voltage over time. Use the curve to make a linear representation from 1 to 100 on a lookup table!

Then if I can get power draw I'll try to incorporate it!

**Reply by thomasklemenc** [reply]
Good luck!

**Reply by DevL0rd** [reply]
> Good luck!

Oh one last thing. Is there a guide somewhere on the process of setting up for development locally and compiling?

**Reply by thomasklemenc** [reply]
There's [this](https://onionui.github.io/docs/dev/setup). though I prefer to use WSL instead of a VM. I wrote a guide on that a long time ago which you can look at [here](https://docs.google.com/document/d/1Uwt7gDVUFDvORWHh_FgLniTDntGuUqKgd4fBpHuOuII)

**Comment by DevL0rd** [comment]
@tGecko here are my findings, and there is a issue with ADC data we are getting. 

#1 i smoothed the incoming ADC value more. Preventing sudden shutdown or battery warning at 30% when the voltage fluctuates

#2 i made the correct voltage curve for li-ion that the miyoo mini has.

just these 2 changes already has made the battery drain nice and linear. https://github.com/OnionUI/Onion/compare/main...DevL0rd:Onion:mmv4-batteryaccuracy




But there is the issue with ADC data.

I tested from 100% to 0%. the data from adc is 100 to 0. but there is 30 - 40 minutes more battery life left after 0 before the 3.0V shutoff of the li-ion. and the curve becomes non linear after around 10% even tho i have the corrected curve lookup table.

Then i realized. the miyoo mini plus and other handhelds using dingux may all us li-po no li-ion.

I think the ADC data we are getting is thinking it's a li-po instead of li-ion and is sending 3.3V as 0 and it should be sending 3v as 0 for li-ion.

After careful testing I can mostly confirm this is the case, but I am not sure how I can resolve it. any input @tGecko ?




Solving this will finally fix the long lasting inaccuracies with the battery as already the new code is far more accurate :)
